### ISSUE TITLE
Modify `output` to follow `expect_out` behaviors

### ISSUE THREAD

**Issue by ryanbmilbourne** [spectcl/spectcl#11]
*Note: Issue edited with updates from comments below.*

In TCL Expect, `expect_out` stores the following:
```
X,string:      //Where X is a number 0...N, one for each capture group in the regular expression.  0 stores the entire string that matched the regex or string provided to the `expect` command
buffer:        //content received from the session until expect matches the input given
spawn_id:      //expect session id

//Optional when using the `--indices` flag
{0..N},start:  //Starting character index for match
{0..N},end:    //Ending character index for match
```
![screen shot 2015-07-25 at 1 03 17 pm](https://cloud.githubusercontent.com/assets/5648689/8891063/99d96efa-32cd-11e5-8cc2-b579a1c695ff.png)
*From Libes, Exploring Expect: A Tcl-based Toolkit for Automating Interactive Programs, p116*

As such, `expect_out(buffer)` would contain the command issued plus any output received before the `expect` command matches:

For the following `expect` block:
```tcl
send { "echo foo\r " }
expect "#" {
   puts $expect_out(buffer)
```

The contents of `expect_out` at the `puts` statement would be:
```
expect_out(0,string):   "#"
expect_out(buffer):     "echo foo
foo
MyPrompt#"
expect_out(spawn_id):   "exp1"
```
So the implication is that it's the `expect` command that is flushing the buffer.  The question is...how closely do we want to mirror this structure?  With my original proposition, we would store the entire session, with output grouped by command.  With the `expect`-style implementation, we only store the buffer of data received between `expect` matches.

We will modify `nexpect` to follow the `expect_out` behavior.  The cache will be stored as a string and will contain any data passed on the session since the last match.

**Comment by ryanbmilbourne** [comment]
After doing a little more digging and reading, `expect_out` actually stores the following:
```
X,string:      //Where X is a number 0...N, one for each capture group in the regular expression.  0 stores the entire string that matched the regex or string provided to the `expect` command
buffer:        //content received from the session until expect matches the input given

//Optional
{0..N},start:  //Starting character index for match
{0..N},end:    //Ending character index for match
```
![screen shot 2015-07-25 at 1 03 17 pm](https://cloud.githubusercontent.com/assets/5648689/8891063/99d96efa-32cd-11e5-8cc2-b579a1c695ff.png)
*From Libes, Expect, p116*

As such, `expect_out(buffer)` would contain the command issued plus any output received before the `expect` command matches:

For the following `expect
``` tcl
send { "echo foo\r " }
expect "#" {
   puts $expect_out(buffer)
```

**Comment by gcochard** [comment]
Well, if the user does want to cache the entire session they can do that. Maybe we should follow it exactly and we can just build a regular expression to leverage it's capture group semantics, which mirror that of TCL.

**Comment by ryanbmilbourne** [comment]
I also favor following TCL's approach.  I'll update the issue.

**Comment by ryanbmilbourne** [comment]
The bulk of this issue was resolved by #17, but there are a few things that remain.  I still need to add the match information to the expect_out object.  I'm thinking that we can vary a little from Expect in this case, and add a match object.  It will be the string that matched, or the RegExp object that triggered the match.
